File: survival-graphics/src/main/java/net/survival/graphics/ModelDisplay.java
package net.survival.graphics;

import java.util.ArrayList;

import org.joml.Matrix4f;

import net.survival.graphics.model.ModelRenderer;
import net.survival.graphics.model.StaticModel;
import net.survival.graphics.opengl.GLMatrixStack;

class ModelDisplay
{
    private final ArrayList<DrawModelCommand> modelsToDraw = new ArrayList<>();
    private final PerspectiveCamera camera;

    private Matrix4f cameraViewMatrix = new Matrix4f();
    private Matrix4f cameraProjectionMatrix = new Matrix4f();

    public ModelDisplay(PerspectiveCamera camera) {
        this.camera = camera;
    }

    public void drawModel(DrawModelCommand model) {
        modelsToDraw.add(model);
    }

    public void display() {
        cameraViewMatrix.identity();
        camera.getViewMatrix(cameraViewMatrix);
        cameraProjectionMatrix.identity();
        camera.getProjectionMatrix(cameraProjectionMatrix);

        GLMatrixStack.setProjectionMatrix(cameraProjectionMatrix);
        GLMatrixStack.push();
        GLMatrixStack.loadIdentity();

        displayPawns(cameraViewMatrix);

        GLMatrixStack.pop();

        modelsToDraw.clear();
    }

    private void displayPawns(Matrix4f viewMatrix) {
        GLMatrixStack.push();
        GLMatrixStack.load(viewMatrix);

        for (var pawn : modelsToDraw)
            displayModel(pawn);

        GLMatrixStack.pop();
    }

    private void displayModel(DrawModelCommand model) {
        if (model.getModelType() != null) {
            GLMatrixStack.push();
            GLMatrixStack.translate(
                    (float) model.getX(),
                    (float) model.getY(),
                    (float) model.getZ());
            GLMatrixStack.rotate((float) model.getYaw(), 0.0f, 1.0f, 0.0f);
            GLMatrixStack.rotate((float) model.getPitch(), 1.0f, 0.0f, 0.0f);
            GLMatrixStack.rotate((float) model.getRoll(), 0.0f, 0.0f, 1.0f);
    
            var displayable = StaticModel.fromModelType(model.getModelType());
            ModelRenderer.displayStaticModel(displayable);
    
            GLMatrixStack.pop();
        }
    }
}

File: survival-graphics/src/main/java/net/survival/graphics/ModelDisplay.kt
package net.survival.graphics

import net.survival.graphics.model.ModelRenderer.displayStaticModel
import net.survival.graphics.model.StaticModel.Companion.fromModelType
import net.survival.graphics.opengl.GLMatrixStack
import org.joml.Matrix4f
import java.util.*

internal class ModelDisplay(
    private val camera: PerspectiveCamera
) {
    private val modelsToDraw = ArrayList<DrawModelCommand>()

    private val cameraViewMatrix = Matrix4f()
    private val cameraProjectionMatrix = Matrix4f()

    fun drawModel(model: DrawModelCommand) {
        modelsToDraw.add(model)
    }

    fun display() {
        cameraViewMatrix.identity()
        camera.getViewMatrix(cameraViewMatrix)
        cameraProjectionMatrix.identity()
        camera.getProjectionMatrix(cameraProjectionMatrix)

        GLMatrixStack.setProjectionMatrix(cameraProjectionMatrix)
        GLMatrixStack.push()
        GLMatrixStack.loadIdentity()

        displayActors(cameraViewMatrix)

        GLMatrixStack.pop()
        modelsToDraw.clear()
    }

    private fun displayActors(viewMatrix: Matrix4f) {
        GLMatrixStack.push()
        GLMatrixStack.load(viewMatrix)
        for (pawn in modelsToDraw) {
            displayModel(pawn)
        }
        GLMatrixStack.pop()
    }

    private fun displayModel(model: DrawModelCommand) {
        GLMatrixStack.push()

        GLMatrixStack.translate(
            model.x.toFloat(),
            model.y.toFloat(),
            model.z.toFloat()
        )
        GLMatrixStack.rotate(model.yaw.toFloat(), 0.0f, 1.0f, 0.0f)
        GLMatrixStack.rotate(model.pitch.toFloat(), 1.0f, 0.0f, 0.0f)
        GLMatrixStack.rotate(model.roll.toFloat(), 0.0f, 0.0f, 1.0f)

        val displayable = fromModelType(model.modelType)
        displayStaticModel(displayable)

        GLMatrixStack.pop()
    }
}